src/wallet/rpcwallet.cpp

```diff
@@ -2397,6 +2399,42 @@ UniValue lockunspent(const JSONRPCRequest& request)
 
         COutPoint outpt(uint256S(txid), nOutput);
 
+        const bool is_locked = pwallet->IsLockedCoin(outpt.hash, outpt.n);
+
+        if (fUnlock && !is_locked) {
+            throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, expected locked output");
+        }
+
+        if (!fUnlock) {
+            if (is_locked) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, output already locked");
+            }
+
+            if (pwallet->IsSpent(outpt.hash, outpt.n)) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, expected unspent output");
+            }
+
+            const auto it = pwallet->mapWallet.find(outpt.hash);
+            if (it == pwallet->mapWallet.end()) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, unknown transaction");
+            }
+
+            const CWalletTx& trans = it->second;
+
+            if (outpt.n >= trans.tx->vout.size()) {
+                throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, vout index out of bounds");
+            }
+        }
+    }
+
+    // Atomically set (un)locked status for the outputs.
+    for (unsigned int idx = 0; idx < outputs.size(); idx++) {
```

Comment:
You could save reparsing the json by saving a vector of `COutPoint`s in the first pass.

Reply:
These loops don't do any json parsing, because the data is already in a UniValue. The `find_value` calls find the desired value in UniValue's internal std::vector of values.

Reply:
You're right. There's no reparsing the JSON, just rereading the string and reconstructing the COutPoint.

I think either way is fine. Whichever you prefer.